README.md
Anzeixer <img src="badge@2x.png" width="130" height="30">
========

Toolkit to Streamline View Definitions for Responsive Web Design


What Is Anzeixer?
-----------------

Anzeixer is a set of CSS and JavaScript snippets that lets you define threshold values of views once, saving time and reducing errors.


How to Use Anzeixer?
--------------------

### Define Views in CSS with @media Rules
Define views that use media queries to apply CSS rules at specific breakpoints and give the views descriptive names. In addition to distinct views for different device widths, you can also create views for portrait and landscape orientations and for various pixel densities.

```css
/**
 * Medium
 */
@media screen and (min-width: 768px) and (max-width: 999px) {

  body:after {
    content: 'm';
    background: #ffe6a0;
  }

}

@media screen and (min-width: 768px) and (max-width: 999px) and (orientation: portrait) {
  body:after { content: 'm portrait'; }
}

@media screen and (min-width: 768px) and (max-width: 999px) and (orientation: landscape) {
  body:after { content: 'm landscape'; }
}
```

### Show the Current View During Development
During development, you need to know that the correct CSS rules are being applied and that the correct JavaScript is being run. Add the "dev" class to the &lt;body&gt; tag to enable an indicator at the bottom left corner which displays the name of the current view.

```html
  <body class="dev">
```

### Query the Current View with JavaScript
You probably need for certain user interface behavior to change depending on the view. For example, you might replace tabs with an accordion on very narrow screens. By using JavaScript to query the current view from CSS instead of defining the threshold values a second time in JavaScript, you reduce complexity and the likelihood of errors.

```js
Anzeixer.getView();      // returns string -- CSS content property of body:after
Anzeixer.getSizeIndex(); // returns integer -- index of current view size

Anzeixer.isXSmall();     // returns boolean -- target smallest devices only (xs)
Anzeixer.isSmall();      // returns boolean -- target all small devices (xs and s)
Anzeixer.isMedium();     // returns boolean -- target medium devices only (m)
Anzeixer.isLarge();      // returns boolean -- target all large devices (l and xl)
Anzeixer.isXLarge();     // returns boolean -- target largest devices only (xl)

Anzeixer.isPortrait();   // returns boolean -- target portrait orientation
Anzeixer.isLandscape();  // returns boolean -- target landscape orientation
```

### Improve Performance with Unified View Change Events
On pages with a lot of JavaScript, having each script listen for the browser's "resize" events individually can impose a significant performance penalty. Many scripts don't need to know the specific viewport size, only which view is in use. Anzeixer can help by triggering a "viewchange" event only when a breakpoint threshold is crossed, letting your scripts ignore "resize" events they don't need.

```js
/**
 * listen to viewchange event
 */
document.addEventListener('viewchange', function(e){
  console.log(e.detail.originalView + ' -> ' + e.detail.currentView);
}, false);
```


Who's behind Anzeixer?
----------------------

Anzeixer is being developed by Zeix's user interface development team.

Anzeixer may be freely used under the MIT license.

Contact: Esther Brunner <esther.brunner@zeix.com>


CHANGELOG.md
# CHANGELOG

## 2.1 (2020-06-15)

- Remove support for jQuery < 3.0
- Update demo for jQuery usage
- Update dev dependencies

## 2.0 (2015-08-14)

- Use Sass to preprocess CSS - breakpoint definitions in _variables.scss
- Use generic view size names: xs, s, m, l, xl
- Add landscape and portrait to view definition strings and introduce isPortrait() and isLandscape() methods
- Fix Firefox print issue
- Update demo files
- Update dev dependencies

## 1.3 (2014-07-14)

- Add event.detail property to jQuery event object
- Update dev dependencies

## 1.2 (2014-03-17)

- New build process based on new grunt packages
- Fix for bug in current Chrome 33
- Fix for Safari 5.1 compatibility

## 1.1 (2013-03-13)

- CustomEvent polyfill for IE9 and IE10
- Updated demo to explain what it does

## 1.0 (2013-09-03)

- Initial release


Gruntfile.js
module.exports = function(grunt) {
  'use strict';

  var sass = require('node-sass');

  grunt.initConfig({
    pkg: grunt.file.readJSON('package.json'),

    uglify: {
      options: {
        // the banner is inserted at the top of the output
        banner: '/*! <%= pkg.name %> <%= pkg.version %> <%= grunt.template.today("dd-mm-yyyy") %> */\n',
        report: 'gzip'
      },
      dist: {
        files: {
          'dist/anzeixer.min.js': ['src/anzeixer.js']
        }
      }
    },

    jshint: {
      options: {jshintrc: '.jshintrc'},
      all: ['Gruntfile.js', 'src/**/*.js']
    },

    copy: {
      demo: {
        src: 'dist/anzeixer.min.js',
        dest: 'demo/js/anzeixer.js',
      },
      dist: {
        src: 'src/anzeixer.js',
        dest: 'dist/anzeixer.js',
      }
    },
    
    sass: {
      options: {
        implementation: sass,
        sourceMap: true,
        outputStyle: 'compressed'
      },
      dist: {
        files: {
          'demo/css/styles.css': 'demo/styles/styles.scss'
        }
      }
    }
  });

  grunt.loadNpmTasks('grunt-contrib-uglify');
  grunt.loadNpmTasks('grunt-contrib-jshint');
  grunt.loadNpmTasks('grunt-contrib-copy');
  grunt.loadNpmTasks('grunt-sass');

  grunt.registerTask('default', ['sass', 'jshint', 'uglify', 'copy']);
  grunt.registerTask('build', ['sass', 'jshint', 'uglify', 'copy']);

};


src/anzeixer.js
/**
 * Custom event polyfill
 * @see https://developer.mozilla.org/en-US/docs/Web/API/CustomEvent?redirectlocale=en-US&redirectslug=Web%2FAPI%2FEvent%2FCustomEvent#Browser_compatibility
 */
try {
  (function() {
    'use strict';

    function CustomEvent(event, params){
      params = params || {bubbles: false, cancelable: false, detail: undefined};
      var evt = document.createEvent('CustomEvent');
      evt.initCustomEvent(event, params.bubbles, params.cancelable, params.detail);
      return evt;
    }

    if (window.CustomEvent === undefined) {
      // Safari 5.1 (and most likely everything below) doesn't have window.CustomEvent
      // function for some reason. It does however create objects with that function as
      // a constructor through the document.createEvent function. So we're poryfilling
      // the polyfill. Yey for JavaScript :|
      var dummy = document.createEvent('CustomEvent');
      CustomEvent.prototype = dummy.constructor.prototype;
    } else {
      CustomEvent.prototype = window.CustomEvent.prototype;
    }

    window.CustomEvent = CustomEvent;
    window.hasCustomEvents = true;
  })();
} catch (ex) {
  // graceful degradation - if there is an unexpected error with the polyfill
  // for custom events, this makes sure that the behavior is still somewhat reasonable
  window.hasCustomEvents = false;
  if (console !== undefined && typeof(console.warn) === 'function') {
    console.warn('Error initializing CustomEvent polyfill - Anzeixer will not raise events');
    console.warn(ex);
  }
}

/**
 * Anzeixer
 *
 * @author Esther Brunner <esther.brunner@zeix.com>
 * @copyright 2013 – 2015 Zeix AG
 * @license Anzeixer.js may be freely distributed under MIT license
 */
var Anzeixer = (function() {
  'use strict';

  /**
   * Get the current view
   * @return {string} string from body:after CSS content property
   */
  var getView = function() {
    try {
      return window.getComputedStyle(document.querySelector('body'), ':after').getPropertyValue('content').replace(/["']/g, '');
    } catch(exception) {
      console.error(exception);
      return 'l landscape'; // assume 1024px wide landscape screen if computed style query fails
    }
  };

  var view = getView(),
      sizes = ['xs', 's', 'm', 'l', 'xl'];

  /**
   * Get the current view and trigger viewchange event if it changed since last query
   *
   * @return {string} string from body:after CSS content property
   */
  function triggerViewChange() {
    var oldView = view;
    view = getView();

    if (oldView !== view && window.hasCustomEvents) {
      var event = new window.CustomEvent('viewchange', {'detail': {
        'originalView': oldView,
        'currentView': view
      }});
      document.dispatchEvent(event);
    }

    return view;
  }

  /**
   * Get index of the current view size
   *
   * @return {integer} zero-based index of sizes array
   */
  var getSizeIndex = function() {
    return sizes.indexOf(view.split(' ')[0]);
  };

  /**
   * Convenience functions for view sizes
   *
   * @return {boolean} whether a size matches the current view
   */
  var isXSmall = function() { return (getSizeIndex() === 0); }; // target smallest devices only
  var isSmall = function() { return (getSizeIndex() < 2); };    // target all small devices
  var isMedium = function() { return (getSizeIndex() === 2); }; // target medium devices only
  var isLarge = function() { return (getSizeIndex() > 2); };    // target all large devices
  var isXLarge = function() { return (getSizeIndex() === 4); }; // target largest devices only

  /**
   * Convenience functions for view orientations
   *
   * @return {boolean} whether a orientation matches the current view
   */
  var isPortrait = function() { return view.split(' ')[1] === 'portrait'; };
  var isLandscape = function() { return view.split(' ')[1] === 'landscape'; };

  // listen to document ready and resize events
  window.addEventListener('DOMContentLoaded', triggerViewChange, false);
  window.addEventListener('resize', triggerViewChange, false);

  // add the detail property to jQuery event object
  if (typeof jQuery !== 'undefined') {
    jQuery.event.props.push('detail');
  }

  return {
    getView: getView,
    getSizeIndex: getSizeIndex,
    isXSmall: isXSmall,
    isSmall: isSmall,
    isMedium: isMedium,
    isLarge: isLarge,
    isXLarge: isXLarge,
    isPortrait: isPortrait,
    isLandscape: isLandscape,

    // @deprecated desktop, tablet, phone view names -- will be removed in Anzeixer 3.0
    isDesktop: isLarge,
    isTablet: isMedium,
    isPhone: isSmall
  };

}());
